core/src/main/java/org/apache/calcite/rel/rules/MinusToDistinctRule.java

```diff
@@ -0,0 +1,184 @@
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Minus;
+import org.apache.calcite.rel.logical.LogicalMinus;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.calcite.util.Util;
+
+import com.google.common.collect.ImmutableList;
+
+import org.immutables.value.Value;
+
+import java.math.BigDecimal;
+
+/**
+ * Planner rule that translates a distinct
+ * {@link org.apache.calcite.rel.core.Minus}
+ * (<code>all</code> = <code>false</code>)
+ * into a group of operators composed of
+ * {@link org.apache.calcite.rel.core.Union},
+ * {@link org.apache.calcite.rel.core.Aggregate},
+ * {@link org.apache.calcite.rel.core.Filter},etc.
+ *
+ * <p>For example, the query plan
+
+ * <blockquote><pre>{@code
+ *  LogicalMinus(all=[false])
+ *    LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2])
+ *      LogicalFilter(condition=[=($7, 10)])
+ *        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ *    LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2])
+ *      LogicalFilter(condition=[=($7, 20)])
+ *        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ * }</pre></blockquote>
+ *
+ * <p> will convert to
+ *
+ * <blockquote><pre>{@code
+ *  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2])
+ *    LogicalFilter(condition=[AND(>($3, 0), =($4, 0))])
+ *      LogicalAggregate(group=[{0, 1, 2}], agg#0=[COUNT() FILTER $3], agg#1=[COUNT() FILTER $4])
+ *        LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], $f3=[=($3, 0)], $f4=[=($3, 1)])
+ *          LogicalUnion(all=[true])
+ *            LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], $f3=[0])
+ *              LogicalFilter(condition=[=($7, 10)])
+ *                LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ *            LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], $f3=[1])
+ *              LogicalFilter(condition=[=($7, 20)])
+ *                LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+ * }</pre></blockquote>
+ *
+ * @see CoreRules#MINUS_TO_DISTINCT
+ */
+@Value.Enclosing
+public class MinusToDistinctRule
+    extends RelRule<MinusToDistinctRule.Config>
+    implements TransformationRule {
+
+  protected MinusToDistinctRule(Config config) {
+    super(config);
+  }
+
+  @Deprecated // to be removed before 2.0
+  public MinusToDistinctRule(Class<? extends Minus> minusClass,
+      RelBuilderFactory relBuilderFactory) {
+    this(MinusToDistinctRule.Config.DEFAULT.withRelBuilderFactory(relBuilderFactory)
+        .as(MinusToDistinctRule.Config.class)
+        .withOperandFor(minusClass));
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Minus minus = call.rel(0);
```

Comment:
```suggestion
    final Minus minus = call.rel(0);
```
Since you have marked all other variables as such, it's sending the message that it will mutate if you don't mark it final too.

Reply:
@asolimando ,thank you for your advice,I will take it.